File: app/src/main/java/com/mvcoding/financius/ui/overview/OverviewPresenter.java
/*
 * Copyright (C) 2015 Mantas Varnagiris.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package com.mvcoding.financius.ui.overview;

import android.support.annotation.NonNull;

import com.mvcoding.financius.ui.ActivityScope;
import com.mvcoding.financius.ui.Presenter;
import com.mvcoding.financius.ui.PresenterView;

import javax.inject.Inject;

import rx.Observable;
import rx.android.view.OnClickEvent;

@ActivityScope class OverviewPresenter extends Presenter<OverviewPresenter.View> {
    @Inject OverviewPresenter() {
    }

    @Override protected void onViewAttached(@NonNull View view) {
        super.onViewAttached(view);
        unsubscribeOnDetach(view.onNewTransactionClick().subscribe(onClickEvent -> view.startNewTransaction()));
    }

    public interface View extends PresenterView {
        @NonNull Observable<OnClickEvent> onNewTransactionClick();

        void startNewTransaction();
    }
}

File: core-app/src/main/kotlin/com/mvcoding/expensius/feature/overview/OverviewPresenter.kt
/*
 * Copyright (C) 2015 Mantas Varnagiris.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package com.mvcoding.expensius.feature.overview

import com.mvcoding.expensius.feature.Filter
import com.mvcoding.expensius.model.ReportPeriod
import com.mvcoding.expensius.service.AppUserService
import com.mvcoding.mvp.Presenter
import org.joda.time.Interval
import rx.Observable
import rx.Observable.combineLatest

class OverviewPresenter(private val appUserService: AppUserService, private val filter: Filter) : Presenter<OverviewPresenter.View>() {

    override fun onViewAttached(view: View) {
        super.onViewAttached(view)

        val filters = filter.filterData().map { it.interval }.distinctUntilChanged()
        val reportPeriods = appUserService.appUser().map { it.settings.reportPeriod }.distinctUntilChanged()
        combineLatest(filters, reportPeriods) { interval, reportPeriod -> interval to reportPeriod }
                .subscribeUntilDetached { view.showInterval(it.second, it.first) }
        view.createTransactionSelects().subscribeUntilDetached { view.displayCreateTransaction() }
        view.transactionsSelects().subscribeUntilDetached { view.displayTransactions() }
        view.tagsSelects().subscribeUntilDetached { view.displayTags() }
        view.settingsSelects().subscribeUntilDetached { view.displaySettings() }
    }

    interface View : Presenter.View {
        fun createTransactionSelects(): Observable<Unit>
        fun transactionsSelects(): Observable<Unit>
        fun tagsSelects(): Observable<Unit>
        fun settingsSelects(): Observable<Unit>

        fun showInterval(reportPeriod: ReportPeriod, interval: Interval)

        fun displayCreateTransaction()
        fun displayTransactions()
        fun displayTags()
        fun displaySettings()
    }
}